### PR TITLE
Add BananaTree

### PR DESCRIPTION
**Pull request by marinaanagno**

### CODE REVIEW

**Comment by marinaanagno** [comment]
@LRacine @samanthaharper

File: doc/examples.xml
Comment (alex-konovalov) [inline doc/examples.xml:316]: Unless this is a standard for Digraphs, no dot is needed for `<Returns>` elements.

File: doc/examples.xml
Comment (alex-konovalov) [inline doc/examples.xml:320]: `n` should be also in `<A>`

**Comment by marinaanagno** [comment]
@alex-konovalov thank you so much for your comments! Fixed them just now! :)

File: doc/examples.xml
Comment (alex-konovalov) [inline doc/examples.xml:1]: Accidental `t` added to this line - will break tests.

File: gap/examples.gi
Comment (flsmith) [inline gap/examples.gi:396]: You could set some other properties here (the obvious one is `IsAcyclicDigraph`)

**Review by flsmith — DISMISSED** [review]
Just a couple of minor things

**Review by alex-konovalov — CHANGES_REQUESTED** [review]
Please fix spurious character in line 1

**Comment by marinaanagno** [comment]
Thank you @wilfwilson! I'm really sorry, I have no idea how I managed to do this, I'll try and fix it now! :)

**Comment by wilfwilson** [comment]
@marinaanagno I just saw that I can delete those files myself through the GitHub interface, so I've done that now. Hopefully this will make the tests pass!

**Review by wilfwilson — APPROVED** [review]

**Comment by marinaanagno** [comment]
@wilfwilson thank you so much! I can't seem to understand why I the files reappeared when I pushed. I deleted them on my branch, but it didn't seem to help :(

File: doc/examples.xml
Comment (wilfwilson) [inline doc/examples.xml:328]:
```suggestion
    a single root vertex that is distinct from all the stars.</Q>
```

File: doc/examples.xml
Comment (wilfwilson) [inline doc/examples.xml:339]: Is this meant to refer to something else? Because this is already the documentation for `BananaTreeGraph`. (Maybe it should be `StarGraph`, for when #424 is merged?)

File: gap/examples.gi
Comment (wilfwilson) [reply gap/examples.gi:396]: @flsmith, non-empty symmetric digraphs are not acyclic!

Perhaps you're thinking of `IsUndirectedTree`.

File: gap/examples.gi
Comment (wilfwilson) [inline gap/examples.gi:397]:
```suggestion
  SetIsUndirectedTree(D, true);
```

File: doc/examples.xml
Comment (wilfwilson) [reply doc/examples.xml]: no, distince → distinct 😆

**Comment by wilfwilson** [comment]
@marinaanagno I'm happy to help with `git` problems if you keep having them 🙂

File: doc/examples.xml
Comment (wilfwilson) [inline doc/examples.xml:318]:
```suggestion
    If <A>m</A> and <A>n</A> are positive integers,
```

File: tst/standard/examples.tst
Comment (wilfwilson) [inline tst/standard/examples.tst:242]:
```suggestion
```
I think you should just delete this test, since it's not testing anything about your new code that the first test on 230 isn't testing.

File: gap/examples.gi
Comment (wilfwilson) [inline gap/examples.gi:405]:
```suggestion
InstallMethod(BananaTreeGraph, "for a function and two pos ints",
```

File: gap/examples.gi
Comment (wilfwilson) [inline gap/examples.gi:401]:
```suggestion
InstallMethod(BananaTreeGraph, "for a function and two pos ints",
```

File: doc/examples.xml
Comment (wilfwilson) [inline doc/examples.xml:320]: I think stating it this way is a bit too tautological, I think something like the following would be better 🙂 
```suggestion
    then this operation returns the <E>banana tree graph</E> 
    with parameters <A>m</A> and <A>n</A>, as defined below. <P/>
```

**Review by wilfwilson — CHANGES_REQUESTED** [review]
Hopefully this will be my last requested changes before merging!

**Comment by wilfwilson** [comment]
> @wilfwilson that would be awesome if you have a second at some point! I'm so sorry for being such a pain today, I don't know how I messed this up :)

Happy to help whenever, just send me a message and we can do some screen sharing if you think that would help.

File: doc/examples.xml
Comment (wilfwilson) [inline doc/examples.xml:339]: Could you please move this to before the paragraph about mutability, and also perhaps change it to a reference to the newly-merged `StarDigraph` instead of complete bipartite digraph? I think a reference to `IsUndirectedTree` would be useful too.
```suggestion
    See also <Ref Oper="StarDigraph"/> and <Ref Prop="IsUndirectedTree"/>.
```
If possible, would you mind being more explicit about the correspondence between the abstract definition and the actual vertices/edges that this function gives you, analogous to what I suggested in https://github.com/digraphs/Digraphs/pull/423/files#r592305981

In particular, I think this documentation should note that vertex `1` is that single root vertex, and that the copies of the stars live on the vertices `[2 .. n+1], [n+2 .. 2*n+1], ...` etc, and also which vertex of each star is connected to `1`.

**Review by wilfwilson — CHANGES_REQUESTED** [review]
This is looking good, thanks for bearing with me with all the changes that I've asked for.

Unfortunately I have a few more! Having reviewed some of the other PRs, I've realised a few things that I want to look out for. I also noticed the following error that your code can run into:
```
gap> BananaTreeGraph(2,1);
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `CompleteBipartiteDigraph' on 3 argument\
s at /Users/Wilf/gap/lib/methsel2.g:249 called from
CompleteBipartiteDigraph( IsMutable, 1, n - 1
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:381 called from
BananaTreeGraphCons( IsMutableDigraph, m, n
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:394 called from
BananaTreeGraphCons( IsImmutableDigraph, m, n
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:403 called from
<function "BananaTreeGraph for a function and two pos ints">( <arguments> )
 called from read-eval loop at *stdin*:4
type 'quit;' to quit to outer loop
```
There are also now merged conflicts between this PR and the `master` branch (because PRs #424 and #425 were merged). Depending on your experience level with git, this might be a bit tricky to resolve. Let me know if you have problems and I'll be happy to help out, or to just resolve the conflicts for you.

**Comment by marinaanagno** [comment]
> This is looking good, thanks for bearing with me with all the changes that I've asked for.
> 
> Unfortunately I have a few more! Having reviewed some of the other PRs, I've realised a few things that I want to look out for. I also noticed the following error that your code can run into:
> 
> ```
> gap> BananaTreeGraph(2,1);
> Error, no method found! For debugging hints type ?Recovery from NoMethodFound
> Error, no 1st choice method found for `CompleteBipartiteDigraph' on 3 argument\
> s at /Users/Wilf/gap/lib/methsel2.g:249 called from
> CompleteBipartiteDigraph( IsMutable, 1, n - 1
>  ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:381 called from
> BananaTreeGraphCons( IsMutableDigraph, m, n
>  ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:394 called from
> BananaTreeGraphCons( IsImmutableDigraph, m, n
>  ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:403 called from
> <function "BananaTreeGraph for a function and two pos ints">( <arguments> )
>  called from read-eval loop at *stdin*:4
> type 'quit;' to quit to outer loop
> ```
> 
> There are also now merged conflicts between this PR and the `master` branch (because PRs #424 and #425 were merged). Depending on your experience level with git, this might be a bit tricky to resolve. Let me know if you have problems and I'll be happy to help out, or to just resolve the conflicts for you.

Thank you so much for all your help @wilfwilson! I will make the changes probably this evening. I have solved a merge conflict manually once or twice, but if there's some other git method that does it differently I'd be very glad to learn about it! :)

File: doc/examples.xml
Comment (marinaanagno) [reply doc/examples.xml:315]: You're absolutely right! I thought that banana tree referred to the way the graph looks (it looks like a banana tree) and didn't consider the possibility that we might mean tree in the graph theoretic sense! :)

**Comment by marinaanagno** [comment]
@wilfwilson git fetch and merge didn't quite work for me (obviously because I did something the wrong way) and I wasn't getting the changes from master. I used git rebase and hopefully I managed to work my way through it. The tests might fail miserably. I will make the rest of the changes now :)

**Review by wilfwilson — CHANGES_REQUESTED** [review]
Thanks for doing those changes. I'd still like for you to explicitly say what the banana tree is, i.e. https://github.com/digraphs/Digraphs/pull/411#discussion_r592391851.

I pushed a couple of tweaks, including I change the variables in the documentation to be `n` and `k`, rather than `m` and `n`. You were quoting Wolfram Alpha, and putting that sentence in `<Q>` tags (i.e. quote marks), but you weren't actually exactly quoting. Now the text is the same.

I've also found a bug. The question is: do you want to support the second argument being 1 or not? If so, then this error shouldn't happen. If not, then it requires a proper error (and documentation) stating that the second argument must be at least 2.
```
gap> BananaTree(1, 1);
Error, the 2nd argument <ran> must be a vertex of the digraph <D> that is the 1st argument, at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:184 called from
DigraphAddEdge( D, edge[1], edge[2]
 ) at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:206 called from
DigraphAddEdge( D, edge ); at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:218 called from
DigraphAddEdges( D, [ [ 1, j * n + 3 ], [ j * n + 3, 1 ] ]
 ); at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:617 called from
BananaTreeCons( IsMutableDigraph, m, n
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:627 called from
BananaTreeCons( IsImmutableDigraph, m, n
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:636 called from
...  at *stdin*:1
type 'quit;' to quit to outer loop
brk> quit;
gap> BananaTree(2, 1);
Error, the 2nd argument <ran> must be a vertex of the digraph <D> that is the 1st argument, at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:184 called from
DigraphAddEdge( D, edge[1], edge[2]
 ) at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:206 called from
DigraphAddEdge( D, edge ); at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:218 called from
DigraphAddEdges( D, [ [ 1, j * n + 3 ], [ j * n + 3, 1 ] ]
 ); at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:617 called from
BananaTreeCons( IsMutableDigraph, m, n
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:627 called from
BananaTreeCons( IsImmutableDigraph, m, n
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:636 called from
...  at *stdin*:1
type 'quit;' to quit to outer loop
brk> quit;
gap> BananaTree(10,1);
Error, the 2nd argument <ran> must be a vertex of the digraph <D> that is the 1st argument, at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:184 called from
DigraphAddEdge( D, edge[1], edge[2]
 ) at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:206 called from
DigraphAddEdge( D, edge ); at /Users/Wilf/gap/pkg/digraphs/gap/oper.gi:218 called from
DigraphAddEdges( D, [ [ 1, j * n + 3 ], [ j * n + 3, 1 ] ]
 ); at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:617 called from
BananaTreeCons( IsMutableDigraph, m, n
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:627 called from
BananaTreeCons( IsImmutableDigraph, m, n
 ) at /Users/Wilf/gap/pkg/digraphs/gap/examples.gi:636 called from
...  at *stdin*:2
type 'quit;' to quit to outer loop
brk>
```

**Comment by marinaanagno** [comment]
Thanks for spotting this @wilfwilson working on it just now! I think the error is what would agree with the theory. As it is defined in BananaTree(n, k) each leaf of S_n is connected to a leaf of each S_k. If k = 1 we don't have a leaf, we have an isolated vertex, so returning an error would seem more reasonable to me for that reason, let me know if you agree. :)
Also if 1 didn't give an error, we would just get a symmetric chain graph, and we can already obtain this graph by taking the symmetric closure of ChainDigraph. Thus, I don't feel that adding 1 as an option would be of any use and since the theory implies that a leaf is connected to a leaf and an isolated vertex is not a leaf, I will proceed by returning an error if you also agree of course. :)

**Comment by wilfwilson** [comment]
Yeah I don't think allowing the second argument to be `1` is useful, so I'm happy to disallow it.

File: doc/examples.xml
Comment (wilfwilson) [inline doc/examples.xml:625]:
```suggestion
    If <A>n</A> and <A>k</A> are positive integers with <A>k</A> greater than <M>1</M>,
```

File: doc/examples.xml
Comment (wilfwilson) [inline doc/examples.xml:649]:
```suggestion
    Specifically, in the resulting digraph, vertex <C>1</C> is the 'root',
    and for each <C>m</C> in <C>[1 .. <A>k</A>]</C>,
    the <C>m</C>th star is on the vertices
    <C>[((m - 1) * n) + 2 .. (m * n) + 1]</C>,
    with the first of these being the 'centre' of the star,
    and the second being the 'leaf' adjacent <C>1</C>. <P/>
```

**Review by wilfwilson — CHANGES_REQUESTED** [review]

**Review by wilfwilson — APPROVED** [review]
I'm happy to merge if the tests pass 🙂

**Pull request merged**
